[PyTorch] Make `List::get() const` match `List::operator[]() const`

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117569
* __->__ #117568

As far as I can tell, `get()` is supposed (and documented) to be the same as a const `operator[]`. We have an efficient implementation for `operator[]`. Let's use it for `get()`.

Differential Revision: [D52809098](https://our.internmc.facebook.com/intern/diff/D52809098/)

swolchok committed:
[PyTorch] Make `List::get() const` match `List::operator[]() const`

As far as I can tell, `get()` is supposed (and documented) to be the same as a const `operator[]`. We have an efficient implementation for `operator[]`. Let's use it for `get()`.

Differential Revision: [D52809098](https://our.internmc.facebook.com/intern/diff/D52809098/)

[ghstack-poisoned]

malfet reviewed (approved):
LGTM

swolchok commented:
Unfortunately, this can break builds of client code, at least in the case of `c10::List<c10::optional<std::string>>`. In that case, the returned type is `c10::optional<std::reference_wrapper<std::string>>`, so you can't do stuff like `list.get(0).value().size()` with the new type (because std::reference_wrapper doesn't have .size()). Looks like we may have accidentally gotten away with this when we changed `List::operator[]`, but there's at least internal stuff that this breaks.

@malfet  am I correct in saying that this is therefore bc-breaking and we can't do it as-is? I can probably just make `List<c10::optional<std::string>>` in particular continue to return a copy to get around that; it looks like the other possible return types are fine.

malfet commented:
@swolchok this is probably fine, as the intent was that `::get() const` should return a read only reference 
Though I'm a bit confused about the example, const reference to element of `c10::List<c10::optional<std::string>>` should be `std::const_reference<c10::optional<std::string>>` shouldn't it?

swolchok committed:
Update on "[PyTorch] Make `List::get() const` match `List::operator[]() const`"


As far as I can tell, `get()` is supposed (and documented) to be the same as a const `operator[]`. We have an efficient implementation for `operator[]`. Let's use it for `get()`.

Differential Revision: [D52809098](https://our.internmc.facebook.com/intern/diff/D52809098/)

[ghstack-poisoned]
